### PR TITLE
feat(pagination_layout): mid-element split for oversized blocks (fulgur-g9e3.1 Phase 3.1)

### PR DESCRIPTION
**Pull request by mitsuru**
## Summary

- Phase 3.1 (`fulgur-g9e3.1`): fragmenter (`pagination_layout`) が body 直下の child の subtree 内も再帰的に split できるように。truly-oversized (`child_h > page_height_px`) ブロックを DOM children に分割し、per-page fragments を emit。
- 新ヘルパ `fragment_block_subtree` を追加。`BlockPageable::find_split_point` の overflow→split パスを mirror。
- skip gate `mid_element_split_skipped` を `forced_break_skipped` に rename + scope 縮小。完全削除と forced-break 対応は follow-up `fulgur-a9qf` (Phase 3.1.5) で実施。

## 背景

Phase 3 epic (`fulgur-g9e3`) は `paginate.rs` 廃止に向けた4段階のうち、3.1 で fragmenter 側の mid-element split を入れる。これまで `BlockPageable::split` に委譲していた "child が strip より大きく、その中の splittable children を切らないとフィットしない" ケース (例: `.huge { break-inside: avoid }` で子に7行 × 80pt) を fragmenter 側で扱えるようにする。

### `crates/fulgur/src/pagination_layout.rs`
- `fragment_block_subtree` を追加。block の children を DOM 順に walk し、各 grandchild について「収まる / 跨ぐ」判定 → 跨ぎ時に page boundary を切り、per-page parent fragment + child fragment を emit。深さは `MAX_DOM_DEPTH` で制限。
- `fragment_pagination_root` の body-direct child loop に `child_h > page_height_px && has_splittable_children` 分岐を追加 → 新ヘルパへ recurse。
- 新 unit test `compare_with_pageable::mid_element_split_keeps_counter_states_in_sync` で row→page placement の regression を cover。

### `crates/fulgur/src/render.rs`
- `mid_element_split_skipped` → `forced_break_skipped` に rename。docstring 更新（残るケースは forced-break recursion 関連のみ、follow-up `fulgur-a9qf` を参照）。

## スコープ外（follow-up: `fulgur-a9qf` / Phase 3.1.5）

advisor 指摘で別 sub-issue に切り出し:

1. **Forced break recursion** — `bare_img_honours_break_before_page` / `pseudo_only_inline_root_honours_break_before_page` / `pseudo_only_list_item_honours_break_before_page` の3件は `column_styles` cascade 漏れ + nested `<li>` の `has_forced_break_below` 未実装が原因。`forced_break_skipped` gate でマスク中。
2. **In-place mid-element split** (`cursor_y + child_h > page_h` かつ `child_h <= page_h`) — parent の CSS-set height と children の sum が乖離するケースでは pre-flight "would actually split" check 無しでは parent fragment height が間違う (例: `<div style=\"height: 600px\">` 内に 30pt の `<h2>` のみ で h=30 を emit、本来 600 であるべき)。`pagination_layout::tests::nested_descendants_inherit_parent_page` で regression が顕在化したため 3.1 では `child_h > page_h_px` gate に絞った。
3. `forced_break_skipped` helper の完全削除 (上記 1, 2 が解消した時点で削除可)。

## Test plan

- [x] `cargo test -p fulgur`: 1111 pass / 0 fail
- [x] `cargo test -p fulgur-vrt` (byte-exact): 31 pass / 0 fail
- [x] `cargo test -p fulgur-cli --test examples_determinism` (byte-exact): 11 pass / 0 fail
- [x] `cargo test -p fulgur-wpt`: 85 pass / 0 fail
- [x] `cargo clippy -p fulgur`: clean
- [x] `cargo fmt --check`: clean
- [x] Target test `tests/break_inside_avoid.rs::avoid_block_taller_than_page_falls_back_to_split` で `assert_pageable_fragmenter_parity` の skip gate が発火しなくなる (paginate==fragmenter==4)

## 関連

- 親 epic: `fulgur-z2mg` (Pageable replacement, geometry-driven render path)
- Phase 3 epic: `fulgur-g9e3` (paginate.rs deletion)
- 前段: PR #281 (`fulgur-q0vb` / Phase 3.0 — spike→fragmenter rename)
- 後段: `fulgur-a9qf` (Phase 3.1.5) → `fulgur-g9e3.2` (Phase 3.2)
- 設計ドキュメント: `docs/plans/2026-04-29-phase3-paginate-deletion.md`

---

<a href="https://app.devin.ai/review/fulgur-rs/fulgur/pull/283" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved pagination handling for oversized block elements. The paginator now better manages page boundaries when splitting large content with splittable children across multiple pages.

* **Tests**
  * Added geometry synchronization smoke test to verify counter-state consistency for pagination edge cases.

### CODE REVIEW

**Comment by mitsuru** [comment]
@coderabbitai review

File: crates/fulgur/src/pagination_layout.rs [hunk 1]
```diff
@@ -722,6 +780,245 @@ fn record_subtree_descendants(
...
+    geometry: &mut PaginationGeometryTable,
+    doc: &BaseDocument,
+    parent_id: usize,
+    parent_w: f32,
+    parent_x_in_body: f32,
+    page_in: u32,
+    cursor_in: f32,
+    page_height_px: f32,
+    depth: usize,
+) -> (u32, f32) {
```
[the remark this answers is not in the document]
Comment (mitsuru) [reply crates/fulgur/src/pagination_layout.rs:856]: ご指摘ありがとうございます。Dismiss というより、Phase 3.1.5 (`fulgur-a9qf`) に統合して deferred します。

理由:
- PR description のとおり Phase 3.1 のスコープは body-direct child の oversized split に限定。再帰経路 (`fragment_block_subtree`) で `running_store` を plumb していないのは事実で、ご指摘どおり nested running marker が in-flow 扱いで `cursor_y` を over-advance します。
- ただし現テスト corpus (gcpm fixtures, `cargo test -p fulgur` 1111/0) では running marker は body-direct のみで、oversized subtree にネストするケースは存在しません。failing test を伴わないため Phase 3.1 で必須とは判断しませんでした。
- `fulgur-a9qf` の notes §5a に scope-add として記録済みです (`bd show fulgur-a9qf` で参照可)。Phase 3.1.5 で in-place mid-element split (#4) と forced-break recursion (1, 2, 3) と合わせて閉じます。
- 63bb9e4 で `fragment_block_subtree` の docstring に "Known gaps deferred to fulgur-a9qf" 節を追加し、この gap が意図的に存在することを明記しました。

Refs: fulgur-a9qf

File: crates/fulgur/src/pagination_layout.rs [hunk 2]
```diff
@@ -722,6 +780,245 @@ fn record_subtree_descendants(
...
+        record_subtree_descendants(
+            geometry,
+            doc,
+            child_id,
+            page_index,
+            cursor_y,
+            child_x_in_body,
+            depth + 1,
+        );
+        cursor_y += child_h;
```
[the remark this answers is not in the document]
Comment (mitsuru) [reply crates/fulgur/src/pagination_layout.rs:1022]: ご指摘ありがとうございます。こちらも Phase 3.1.5 (`fulgur-a9qf`) に統合して deferred します。

理由:
- ご指摘の divergence は事実です。body-level walker は `collect_inline_line_metrics` + `fragment_inline_root` で line edge split しますが、`fragment_block_subtree` の再帰経路は DOM-child block split にフォールバックしています。
- 現テスト corpus には「巨大 div の中にページを越える `<p>`」を含む fixture がなく、`cargo test -p fulgur` (1111/0) では顕在化していません。Phase 3.2 (`fulgur-g9e3.2`、`paginate()` 廃止 + paragraph 経路置換) との接続点でもあるため、3.1.5 着手時に 3.2 に統合するか改めて判断します。
- `fulgur-a9qf` の notes §5b に scope-add として記録済みです。
- 63bb9e4 で `fragment_block_subtree` の docstring に明記しました。

Refs: fulgur-a9qf, fulgur-g9e3.2

File: crates/fulgur/src/pagination_layout.rs [hunk 3]
```diff
@@ -722,6 +780,245 @@ fn record_subtree_descendants(
...
+                );
+                page_index = np;
+                cursor_y = nc;
+                // The recursion may have moved us off `page_start_y`'s
+                // page; restart the parent fragment on whatever page
+                // we're on now.
+                if page_index != page_in || nc < page_start_y {
+                    page_start_y = 0.0;
+                }
+                prev_bottom_in_parent = this_top_in_parent + child_h;
```
[the remark this answers is not in the document]
Comment (mitsuru) [reply crates/fulgur/src/pagination_layout.rs:990]: ご指摘ありがとうございます。こちらも Phase 3.1.5 (`fulgur-a9qf`) に統合して deferred します。Critical 表記とのこと、慎重に検討しました。

理由:
- ご指摘どおり、再帰呼び出しが複数ページ跨ぐとき (`np > page_in + 1`)、`page_start_y = 0` の reset しかしておらず、ancestor (現在の `parent_id`) の per-page fragment を中間ページ (`page_in+1..np`) に emit していないのは事実です。
- 影響範囲: counter / string-set / bookmark op が **2段ネストの oversized 親に attach** された場合のみ parity 破綻します。現テスト (`mid_element_split_keeps_counter_states_in_sync` 含む 1111/0) は子側に counter op を attach しており、この parity 破綻は顕在化していません。
- 本質的に既に scope に入っている in-place mid-element split (#4) と同じ「親 fragment per-page 整合性」カテゴリで、修正は前後関係を含めて一緒に書く方が綺麗になります。`fulgur-a9qf` の notes §5c に scope-add として記録、`fulgur-7hf5` (3.1.5c: in-place mid-element split + remove forced_break_skipped) のスコープに統合する方針です。
- 63bb9e4 で `fragment_block_subtree` の docstring に "Multi-page recursive traversal does not emit per-page parent fragments for intermediate pages" として明記しました。

Phase 3.1.5 着手時に 2段ネストの oversized + ancestor counter op の regression test を新設して fix します。

Refs: fulgur-a9qf, fulgur-7hf5
